packet.go

```diff
@@ -499,34 +513,47 @@ func (p *Packet) MarshalTo(buf []byte) (n int, err error) {
 		return 0, err
 	}
 
+	return marshalPayloadAndPaddingTo(buf, n, &p.Header, p.Payload)
+}
+
+func marshalPayloadAndPaddingTo(buf []byte, offset int, header *Header, payload []byte) (n int, err error) {
 	// Make sure the buffer is large enough to hold the packet.
-	if n+len(p.Payload)+int(p.PaddingSize) > len(buf) {
+	if offset+len(payload)+int(header.PaddingSize) > len(buf) {
 		return 0, io.ErrShortBuffer
 	}
 
-	m := copy(buf[n:], p.Payload)
+	m := copy(buf[offset:], payload)
 
-	if p.Header.Padding {
-		buf[n+m+int(p.PaddingSize-1)] = p.PaddingSize
+	if header.Padding {
+		buf[offset+m+int(header.PaddingSize-1)] = header.PaddingSize
 	}
 
-	return n + m + int(p.PaddingSize), nil
+	return offset + m + int(header.PaddingSize), nil
 }
 
 // MarshalSize returns the size of the packet once marshaled.
 func (p Packet) MarshalSize() int {
-	return p.Header.MarshalSize() + len(p.Payload) + int(p.PaddingSize)
+	if p.PaddingSize > 0 && p.Header.PaddingSize == 0 {
+		p.Header.PaddingSize = p.PaddingSize
```

Comment:
The function name looks like an immutable method but have a side effect. (`Clone()` as well)
I think it's reasonable for now but maybe better to note this as a godoc comment.

Reply:
Thanks. I have updated code to remove this side effect.